CDK-1000: Kite/SparkSQL integration

dgreco opened the pull request.

tomwhite commented on kite-data/kite-data-spark/src/main/scala/org/kitesdk/data/spark/KiteDatasetReader.scala:
```diff
@@ -0,0 +1,33 @@
+
+object KiteDatasetReader {
+
+  def readAsDataFrame(sqlContext: SQLContext,
```
Rename KiteDatasetLoader/loadAsDataFrame for symmetry with KiteDatasetSaver/saveAsKiteDataset.

dgreco replied:
Yes, it's better

tomwhite commented on kite-data/kite-data-spark/src/main/scala/org/kitesdk/data/spark/KiteDatasetSaver.scala:
```diff
@@ -0,0 +1,166 @@
+    */
+    val schema = partitionStrategy.fold(dataFrame.schema)(_ =>
+      StructType(dataFrame.schema.iterator.map(field =>
+        if (field.nullable) field.copy(nullable = false) else field).toArray))
```
Did https://issues.cloudera.org/browse/CDK-986 not remove the need for this?

dgreco replied:
Nope

tomwhite commented on kite-data/kite-data-spark/src/main/scala/org/kitesdk/data/spark/KiteDatasetSaver.scala:
```diff
@@ -0,0 +1,166 @@
+                                format: Format,
+                                compressionType: CompressionType,
+                                partitionStrategy: Option[PartitionStrategy],
+                                columnMapping: Option[ColumnMapping],
```
These parameters should be wrapped up in a DatasetDescriptor. Also we should think about whether we need a DatasetDescriptor at all, or whether it is optional, so that existing datasets can be used. CrunchDatasets appends data, and has a handleExisting check (see DatasetTarget).

dgreco replied:
I was thinking to pass a descriptor instead, The problem I found is that I have to specify everything in the descriptor including the schema, but the schema is generated internally from the passed data frame.

For writing I'm using the kite output format, so if it's possible to append using the kite output format I could do it as well.

dgreco replied:
An alternative could to specify a specific SparkDatasetDescriptor that allows specifying the dataframe which is then used internally for generating the schema, something like this:
```scala
val descriptor = new SparkDatasetDescriptor.Builder().dataFrame(teenagers).format(format).compressionType(CompressionType.Snappy).build()

val dataset = KiteDatasetSaver.saveAsKiteDataset(descriptor, datasetURI)
```
What do you think?

tomwhite replied:
Yes, this is better.

tomwhite commented on kite-data/kite-data-spark/src/main/scala/org/kitesdk/data/spark/KiteDatasetSaver.scala:
```diff
@@ -0,0 +1,166 @@
+                                compressionType: CompressionType,
+                                partitionStrategy: Option[PartitionStrategy],
+                                columnMapping: Option[ColumnMapping],
+                                job: Option[Job] = None): Dataset[Record] = {
```
We shouldn't need to expose Job here.

dgreco replied:
I'm passing an Option[Job] so if it's defined I'll use the passed job otherwise I would create a Job instance from the external context. In some tests I need to pass a job, because it's used to pass the configuration associated to the base mini cluster. That parameter is None by default so it's not used normally.

tomwhite replied:
Kite has a slightly different way of passing configuration information, via the DefaultConfiguration. You should be able to use that from tests, and so not pass it here.

dgreco replied:
ok, good suggestion. I'll fix that.

tomwhite commented on kite-data/kite-data-spark/src/test/scala/org/kitesdk/data/spark/SparkKiteSpec.scala:
```diff
@@ -0,0 +1,305 @@
+      val conf = new SparkConf().
+        setAppName("spark-kite-spec-test").
+        set("spark.serializer", "org.apache.spark.serializer.KryoSerializer").
+        setMaster("local[16]")
```
Can this just be `local`?

tomwhite commented on kite-data/kite-data-spark/src/test/scala/org/kitesdk/data/spark/SparkKiteSpec.scala:
```diff
@@ -0,0 +1,305 @@
+  }
+
+  "Spark" must {
+    "be able to create a SchemaRDD/Dataframe from a kite avro dataset" in {
```
This test is basically the same as the previous one, except it uses Avro - can you factor out into a common method?

dgreco replied:
I'll do it.

tomwhite commented on kite-data/kite-data-spark/src/test/scala/org/kitesdk/data/spark/SparkKiteSpec.scala:
```diff
@@ -0,0 +1,305 @@
+      val sqlContext = new SQLContext(sparkContext)
+      import sqlContext.implicits._
+
+      val datasetURI = URIBuilder.build(s"repo:file:////${System.getProperty("user.dir")}/tmp", "test", "persons")
```
Create temp paths under target.

tomwhite commented:
Thanks for the update @dgreco! I added some more review comments. Have you looked at why the Travis tests are failing?

It might be good to get more eyes on this - @joey wrote the original Spark integration so it would be good to have his take on this change.

You don't need to open new PRs for each change BTW. It's best to have one so that the discussion is in one place.

dgreco commented:
The spark specific tests are passing now, it seems now that some of the tests are failing because of lacking of resources

tomwhite commented:
One of the test failures is due to multiple Spark contexts running at once. Can you try setting spark.driver.allowMultipleContexts to true?

dgreco commented:
Tom, I committed other stuff, commits:  8d79a53 and  53aa87e 
I'm still learning how to use PRs, did you check what I did? I changed the apis, I tried to reuse the 
concept of dataset descriptor to be passed to the save method, instead of passing multiple parameters.
I also added a method for appending/overwriting to existing datasets. It looks nicer to me, clean apis.
And less code.

tomwhite commented:
Thanks for the latest changes @dgreco. It looks like the Spark tests are failing with NPEs:

https://travis-ci.org/kite-sdk/kite/jobs/63672318

Also Flume, possibly due to changes in the library versions:

https://travis-ci.org/kite-sdk/kite/jobs/63672316

Do these pass locally?

dgreco commented:
Hi Tom,
yes they are passing locally, I'm trying with the same commands that travis is using.
It looks like that travis is exhausting resources.

dgreco commented:
Hi Tom,
finally I made the tests running on Travis, the problem was related to the spark context management when there are multiple spark tests. I refactored the tests to be sure that only one spark context is running at same moment.

dgreco commented:
I close this PR since I'm restructuring my git repo
